File: R/moveit.R
#' Authenticate User
#'
#' @param baseUrl Base URL for Move It server
#' @param payload Authentication header
#'
#' @return List of auth tokens for MoveIt access
#' @export
#'
#' @examples
#' \dontrun{
#' tokens <- authMoveIt("someurl.com", "grant_type=password&username=USERNAME&password=PASSWORD")
#' }
authMoveIt <- function(baseUrl, payload) {
  # Check function dependancies
  if (!requireNamespace("httr", quietly = TRUE)) {
    stop("Package \"httr\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  # Build URL
  url <- paste0("https://moveit.", baseUrl, "/api/v1/token")
  # Post Auth
  g <- httr::POST(url, body = payload, httr::add_headers('Content-Type' = 'application/json'))
  # Get tokens
  tokens <- httr::content(g)
  # Return auth token list
  return(tokens)
}

#' Function to download a file from the FTP server
#'
#' @param baseUrl Base URL for Move It server
#' @param tokens List of auth tokens for MoveIt access
#' @param id file id
#' @param fileType csv or xlsx
#'
#' @return A dataframe of the identified file
#' @export
#'
#' @examples
#' \dontrun{
#' df <- readMoveItFile("someurl.com", 626235843, tokens, "csv")
#' }
readMoveItFile <- function(baseUrl, tokens, id, fileType = "csv") {
  # Check function dependancies
  if (!requireNamespace("httr", quietly = TRUE)) {
    stop("Package \"httr\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  if (fileType == "csv") {
    if (!requireNamespace("readr", quietly = TRUE)) {
      stop("Package \"readr\" needed for this function to work. Please install it.",
           call. = FALSE)
    }
  } else if (fileType == "excel") {
    if (!requireNamespace("readxl", quietly = TRUE)) {
      stop("Package \"readxl\" needed for this function to work. Please install it.",
           call. = FALSE)
    }
  }

  # Create Temp file for download
  tmp <- tempfile()
  token <- tokens$access_token
  # Build file URL
  url <- paste0("https://moveit.", baseUrl, "/api/v1/files/", id, "/download")
  # Request
  if (fileType == "excel") {
    cType <- "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
  } else {
    cType  <- "text/csv"
  }
  g <- httr::GET(url,
                 httr::add_headers(Authorization = paste("Bearer", token)),
                 httr::content_type(cType),
                 httr::write_disk(tmp))

  # Read tmp file by file type
  if (fileType == "csv") {
    data <- readr::read_csv(tmp, guess_max = 5000)
  } else if (fileType == "excel") {
    data <- readxl::read_excel(tmp)
  }
  # Delete Temp File
  file.remove(tmp)
  # Return data to user
  return(data)
}

#' Available Files
#'
#' @param baseUrl Base URL for Move It server
#' @param tokens List of auth tokens for MoveIt access
#'
#' @return Dataframe of available files and their details
#' @export
#'
#' @examples
#' \dontrun{
#' files <- availableFiles("someurl.com", tokens)
#' }
availableFiles <- function(baseUrl, tokens) {
  # Check function dependancies
  if (!requireNamespace("httr", quietly = TRUE)) {
    stop("Package \"httr\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("Package \"jsonlite\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  # Read access Token
  token <- paste("Bearer", tokens$access_token)
  # Build URL
  url <- paste0("https://moveit.", baseUrl, "/api/v1/files")
  # Send Get request
  g <- httr::GET(url,
                 httr::add_headers(Authorization = token))
  # Files list
  files <- jsonlite::fromJSON(httr::content(g, "text"))

  # Files dataframe
  items <- files$items
  # Page Number & total pages for while loop
  page <- as.numeric(files$paging$page)
  totalPages <- as.numeric(files$paging$totalPages)

  # Iterate over additional pages
  while (page != totalPages) {
    # Dependency Check
    if (!requireNamespace("dplyr", quietly = TRUE)) {
      stop("Package \"dplyr\" needed for this function to work. Please install it.",
           call. = FALSE)
    }
    # Next Page number
    nextPage <- page + 1
    # Build URL
    urlPage <- paste0(url, "?page=", nextPage)

    # Get page of files
    g <- httr::GET(urlPage,
                   httr::add_headers(Authorization = token))

    # Files list
    files <- jsonlite::fromJSON(httr::content(g, "text"))

    # Bind to previous pages
    items <- dplyr::bind_rows(items, files$items)

    # Get current page
    page <- as.numeric(files$paging$page)
    # Check totalPages
    totalPages <- as.numeric(files$paging$totalPages)
  }
  return(items)
}

#' List Available Folder
#'
#' @param baseUrl Base URL for Move It server
#' @param tokens List of auth tokens for MoveIt access
#'
#' @return Dataframe of available folders and their details
#' @export
#'
#' @examples
#' \dontrun{
#' folders <- availableFolders("someurl.com", tokens)
#' }
availableFolders <- function(baseUrl, tokens) {
  # Check dependancies
  if (!requireNamespace("httr", quietly = TRUE)) {
    stop("Package \"httr\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("Package \"jsonlite\" needed for this function to work. Please install it.",
         call. = FALSE)
  }

  # Load Auth token
  token <- paste("Bearer", tokens$access_token)
  # Build URL
  url <- paste0("https://moveit.", baseUrl, "/api/v1/folders")

  # Get request for folders
  g <- httr::GET(url,
                 httr::add_headers(Authorization = token))

  # List of folders
  folders <- jsonlite::fromJSON(httr::content(g, "text"), flatten = TRUE)

  # Folder details as dataframe
  items <- folders$items

  # Page Number & total pages for while loop
  page <- as.numeric(folders$paging$page)
  totalPages <- as.numeric(folders$paging$totalPages)

  # Iterate over additional pages
  while (page != totalPages) {
    # Dependency Check
    if (!requireNamespace("dplyr", quietly = TRUE)) {
      stop("Package \"dplyr\" needed for this function to work. Please install it.",
           call. = FALSE)
    }
    # Next Page number
    nextPage <- page + 1
    # Build URL
    urlPage <- paste0(url, "?page=", nextPage)

    # Get page of files
    g <- httr::GET(urlPage,
                   httr::add_headers(Authorization = token))
    # Folders list
    folders <- jsonlite::fromJSON(httr::content(g, "text"), flatten = TRUE)

    # Bind to previous pages
    items <- dplyr::bind_rows(items, folders$items)
    # Get current page and total pages
    page <- folders$paging$page
    totalPages <- folders$paging$totalPages
  }
  return(items)
}

#' Upload file to move it server
#'
#' @param baseUrl Base URL for Move It server
#' @param tokens List of auth tokens for MoveIt access
#' @param folderId ID for destination folder
#' @param filePath Path to file for upload
#' @param fileType POST file type ie "text/csv"
#' @param chunked TRUE or FALSE value, TRUE forces chunked upload. Function will automatically chunk if file is too big
#'
#' @return POST content
#' @export
#'
#' @examples
#' \dontrun{
#' uploadMoveItFile("someurl.com", tokens, 2346247, "some.csv", "text/csv")
#' }
uploadMoveItFile <- function(baseUrl, tokens, folderId, filePath, fileType, chunked=FALSE) {
  # Check dependency
  if (!requireNamespace("httr", quietly = TRUE)) {
    stop("Package \"httr\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  # Fix Chunked on Linux
  if(missing(chunked)) {
    chunked <- FALSE
  }

  # Load Auth token
  token <- paste("Bearer", tokens$access_token)

  # Build URL
  url <- paste0("https://moveit.", baseUrl, "/api/v1/folders/", folderId, "/files")

  size <- file.size(filePath)

  if (size >= 40000000 | chunked) {
    # Send Request
    return <- httr::POST(url = url,
               httr::add_headers(Authorization = token,
                                 accept = "application/json",
                                 `Content-Type` = "multipart/form-data",
                                 `Transfer-Encoding` = "chunked"),
               body = list(file =  httr::upload_file(filePath, fileType))
               )
  } else {
    return <- httr::POST(url = url,
                         httr::add_headers(Authorization = token,
                                           accept = "application/json",
                                           `Content-Type` = "multipart/form-data"),
                         body = list(file =  httr::upload_file(filePath, fileType))
    )
  }
  if (!return$status_code %in% c(201, 200)) {
    stop(return$status_code)
  }
}

File: Python/src/moveitAPI/moveit.py
import requests
import pandas as pd
import tempfile
import os
import json


def auth_move_it(base_url, payload):
    # Build URL
    url = f"https://moveit.{base_url}/api/v1/token"
    # Post Auth
    response = requests.post(url, data=payload, headers={'Content-Type': 'application/json'})
    try:
        response.raise_for_status()
    except requests.exceptions.HTTPError as e:
        return "Error: " + str(e)
    # Get tokens
    tokens = response.json()
    # Return auth token list
    return tokens


def read_move_it_file(base_url, tokens, file_id, file_type="csv", sheet_download=None):
    # Create Temp file for download
    tmp = tempfile.NamedTemporaryFile(delete=False)
    token = tokens['access_token']
    # Build file URL
    url = f"https://moveit.{base_url}/api/v1/files/{file_id}/download"
    # Request
    if file_type == "excel":
        c_type = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
    else:
        c_type = "text/csv"

    headers = {'Authorization': f'Bearer {token}'}
    response = requests.get(url, headers=headers, stream=True)

    with open(tmp.name, 'wb') as f:
        f.write(response.content)

    # Read tmp file by file type
    if file_type == "csv":
        data = pd.read_csv(tmp.name, low_memory=False)
    elif file_type == "txt":
        data = pd.read_csv(tmp.name, sep='\t', low_memory=False)
    elif file_type == "excel":
        data = pd.read_excel(tmp.name, sheet_name=sheet_download)

    # Delete Temp File
    os.remove(tmp.name)
    # Return data to user
    return data


def available_files(base_url, tokens):
    # Read access Token
    token = f"Bearer {tokens['access_token']}"
    # Build URL
    url = f"https://moveit.{base_url}/api/v1/files"
    # Send Get request
    response = requests.get(url, headers={"Authorization": token})
    # Files list
    files = response.json()

    # Files dataframe
    items = files['items']
    # Page Number & total pages for while loop
    page = int(files['paging']['page'])
    total_pages = int(files['paging']['totalPages'])

    # Iterate over additional pages
    while page != total_pages:
        # Next Page number
        next_page = page + 1
        # Build URL
        url_page = f"{url}?page={next_page}"

        # Get page of files
        response = requests.get(url_page, headers={"Authorization": token})

        # Files list
        files = response.json()

        # Bind to previous pages
        items.extend(files['items'])

        # Get current page
        page = int(files['paging']['page'])
        # Check total_pages
        total_pages = int(files['paging']['totalPages'])

    return items


def available_folders(base_url, tokens):
    # Load Auth token
    token = f"Bearer {tokens['access_token']}"
    # Build URL
    url = f"https://moveit.{base_url}/api/v1/folders"

    # Get request for folders
    response = requests.get(url, headers={'Authorization': token})

    # List of folders
    folders = response.json()

    # Folder details as list of dictionaries
    items = folders['items']

    # Page Number & total pages for while loop
    page = int(folders['paging']['page'])
    total_pages = int(folders['paging']['totalPages'])

    # Iterate over additional pages
    while page != total_pages:
        # Next Page number
        next_page = page + 1
        # Build URL
        url_page = f"{url}?page={next_page}"

        # Get page of files
        response = requests.get(url_page, headers={'Authorization': token})
        folders = response.json()

        # Bind to previous pages
        items.extend(folders['items'])
        # Get current page and total pages
        page = folders['paging']['page']
        total_pages = folders['paging']['totalPages']

    return items


def upload_move_it_file(base_url, tokens, folder_id, file_path, file_type, chunked=False):
    # Check dependency
    try:
        import requests
    except ImportError:
        raise ImportError("Package 'requests' needed for this function to work. Please install it.")

    # Fix Chunked on Linux
    if chunked is None:
        chunked = False

    # Load Auth token
    token = f"Bearer {tokens['access_token']}"

    # Build URL
    url = f"https://moveit.{base_url}/api/v1/folders/{folder_id}/files"

    size = os.path.getsize(file_path)

    headers = {
        "Authorization": token,
        "accept": "application/json",
        "Content-Type": "multipart/form-data"
    }

    if size >= 40000000 or chunked:
        headers["Transfer-Encoding"] = "chunked"

    with open(file_path, 'rb') as f:
        files = {'file': (os.path.basename(file_path), f, file_type)}
        response = requests.post(url, headers=headers, files=files)

    if response.status_code not in [201, 200]:
        raise Exception(response.status_code)


